companies/views.py

```diff
@@ -324,71 +320,56 @@ def companies_list(request, year):
         else:
             signatures[signature.company].append(signature)
 
-    exhibitors = [
-        x.company
-        for x in Exhibitor.objects.select_related("company").filter(fair=exhibitor_fair)
-    ]
+    # Pagination variables
+    COMPANIES_PER_PAGE = 10
+    page_number = int(request.GET.get("page") or 1)
+
+    # SQL level filtering
+    total_companies = Company.objects.prefetch_related("groups", "companycontact_set")
+    filtered_companies = total_companies
+    exhibitors = Exhibitor.objects.select_related("company").filter(fair=exhibitor_fair)
+    if has_filtering:
+        if form.cleaned_data["exhibitors"] == "YES":
+            filtered_companies = filtered_companies.filter(pk__in=exhibitors)
+        elif form.cleaned_data["exhibitors"] == "NO":
+            filtered_companies = filtered_companies.exclude(pk__in=exhibitors)
+        if len(form.cleaned_data["contracts_positive"]) != 0:
+            contract_signing_companies = (
+                SignupLog.objects.select_related("company")
+                .select_related("contract")
+                .filter(contract__in=form.cleaned_data["contracts_positive"])
+            )
+            filtered_companies = filtered_companies.filter(
+                pk__in=contract_signing_companies.values_list("company")
+            )
+        if len(form.cleaned_data["contracts_negative"]) != 0:
+            contract_signing_companies = (
+                SignupLog.objects.select_related("company")
+                .select_related("contract")
+                .filter(contract__in=form.cleaned_data["contracts_negative"])
+            )
+            filtered_companies = filtered_companies.exclude(
+                pk__in=contract_signing_companies.values_list("company")
+            )
+        if len(form.cleaned_data["users"]) != 0:
+            companies_with_sought_responsibles = (
+                CompanyCustomerResponsible.objects.select_related("company")
+                .select_related("group")
+                .filter(group__fair=fair)
+                .prefetch_related("users")
+                .filter(users__in=form.cleaned_data["users"])
+                .values_list("company")
+            )
+            filtered_companies = filtered_companies.filter(
+                pk__in=companies_with_sought_responsibles
+            )
 
     companies_modified = []
-
-    for company in companies:
+    for company in filtered_companies[
```

Comment:
This for loop header is quite cluttered, could we move out the

`(page_number - 1) * COMPANIES_PER_PAGE : page_number * COMPANIES_PER_PAGE`

to its own variable with a logical name?